python-modules/cis_profile/cis_profile/profile.py

```diff
@@ -100,6 +100,66 @@ def get_profile_from_file(self, user_structure_json_path):
             path = user_structure_json_path
         return DotDict(json.load(open(path)))
 
+    def merge(self, user_to_merge_in, publisher, level=None, _internal_level=None):
+        """
+        Recursively merge user attributes:
+        Merge a User object with another User object. This will override all fields from the current user by non-null
+        (non-None) fields from `user_to_merge_in` which have a publisher set to `publisher`.
+        Ex:
+        u_orig = User()
+        u_patch = User()
+        u_patch.timezone.value = "GMT -07:00 America/Los Angeles"
+
+        u_orig.merge(u_patch, publisher="hris")
+        print(u_orig.timezone.value)
+        # >>> GMT -07:00 America/Los Angeles
+
+        This is useful when updating existing users, rather than manually matching fields. Passing the publisher ensures
+        that we only patch the fields that are allowed for this publisher, as a safeguard, which ensures validation will
+        work
+
+        @user_to_merge_in User object the user to merge into the current user object
+        @publisher str the publisher name that attempts to merge data in. any data from `user_to_merge_in` that is not
+        originating from this publisher will be ignored
+        @level dict of an attribute. This can be user_to_merge_in.__dict__ for the top level (recurses through all
+        attributes). The level MUST be from user_to_merge_in, not from the original/current user.
+        @_internal_level str attribute name  of previous level attribute from the current user. Used internally.
+
+        This function always returns None and is always successful.
+        """
+        tomerge = []
+        # Default to top level
+        if level is None:
+            level = user_to_merge_in.__dict__
+
```

Comment:
```python
if _intenal_level is None:
    _inteanl_level = self.__dict__
```

Reply:
@fiji-flo can you write some words about why this is needed?

Reply:
(or is that just so that `level` and `_internal_level` are both handled the same way? I don't think that changes much though, right?)

Reply:
Wow 2 typos in 2 lines :P

```python
if _internal_level is None:
    _internal_level = self.__dict__
```
would do the initialization of `_internal_level` when first called on the top level object. In all recursions `_internal_level` will be set already.
So yes this is to handle `level` and `_internal_level` the same way. What changes is: it replaces  [line 149](https://github.com/mozilla-iam/cis/pull/315/files#diff-6c6103298575e3cf2052ba6a215d0060R149):
and would allow merging of fields deeper than 2 levels.

This only works with my other comment in place.

Reply:
side note: we never have more than 2 level

Reply:
Yes :P
But I would vote for either doing this iterative and only support 2 levels or recursive and support >2 levels.